apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts

```diff
@@ -52,22 +96,86 @@ router.post(
     if (!config.openAiApiKey || !config.openAiOrganization) {
       throw new error.HttpStatusError(403, 'Not implemented (feature not available)');
     }
+
     const client = new OpenAI({
       apiKey: config.openAiApiKey,
       organization: config.openAiOrganization,
     });
 
     if (req.body.__action === 'generate_question') {
-      const jobSequenceId = await generateQuestion(
+      const result = await generateQuestion(
         client,
         res.locals.course ? res.locals.course.id : undefined,
         res.locals.authn_user.user_id,
         req.body.prompt,
       );
-      res.redirect(res.locals.urlPrefix + '/jobSequence/' + jobSequenceId);
+
+      const data = result.jobResult;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, result.jobSequenceId, res.locals),
+        );
+      } else {
+        throw new error.HttpStatusError(500, `Job sequence ${result.jobSequenceId} failed.`);
+      }
     } else if (req.body.__action === 'sync_context_documents') {
       const jobSequenceId = await syncContextDocuments(client, res.locals.authn_user.user_id);
       res.redirect('/pl/jobSequence/' + jobSequenceId);
+    } else if (req.body.__action === 'regenerate_question') {
+      const genJobs = await selectJobsByJobSequenceId(req.body.unsafe_sequence_job_id);
+      if (
+        genJobs.length !== 1 ||
+        !genJobs[0]?.course_id ||
+        !idsEqual(genJobs[0]?.course_id, res.locals.course.id)
+      ) {
+        throw new error.HttpStatusError(
+          403,
+          `Job sequence ${req.body.unsafe_sequence_job_id} not found.`,
+        );
+      }
+
+      const result = await regenerateQuestion(
+        client,
+        res.locals?.course?.course_id,
+        res.locals.authn_user.user_id,
+        genJobs[0]?.data?.prompt,
+        req.body.prompt,
+        genJobs[0]?.data?.html,
+        genJobs[0]?.data?.python,
+      );
+
+      const data = result.jobResult;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, result.jobSequenceId, res.locals),
+        );
+      } else {
+        res.redirect('/pl/jobSequence/' + result.jobSequenceId);
+      }
```

Comment:
Same comment as above, if you're so inclined.

Reply:
Changed in 6919d9f4829f12ab87c64d40f2dccf405a98d563.